// piral-base/lib/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startLoadingPilets = void 0;
const utils_1 = require("./utils");
const strategies_1 = require("./strategies");
function startLoadingPilets(options) {
    const state = {
        loaded: false,
        pilets: [],
        error: undefined,
    };
    const notifiers = [];
    const call = (notifier) => notifier(state.error, state.pilets, state.loaded);
    const notify = () => notifiers.forEach(call);
    const setPilets = (error, pilets) => {
        state.error = error;
        state.pilets = pilets;
        notify();
    };
    const setLoaded = () => {
        state.loaded = true;
        notify();
    };
    const { strategy = strategies_1.standardStrategy } = options;
    strategy(options, setPilets).then(setLoaded, setLoaded);
    return {
        connect(notifier) {
            if ((0, utils_1.isfunc)(notifier)) {
                notifiers.push(notifier);
                call(notifier);
            }
        },
        disconnect(notifier) {
            const index = notifiers.indexOf(notifier);
            index !== -1 && notifiers.splice(index, 1);
        },
    };
}
exports.startLoadingPilets = startLoadingPilets;
