// piral-core/lib/components/ForeignComponentContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForeignComponentContainer = void 0;
const React = require("react");
const piral_base_1 = require("piral-base");
class ForeignComponentContainer extends React.Component {
    constructor() {
        super(...arguments);
        this.local = {};
        this.handler = (ev) => {
            const { innerProps } = this.props;
            ev.stopPropagation();
            innerProps.piral.renderHtmlExtension(ev.detail.target, ev.detail.props);
        };
        this.setNode = (node) => {
            this.current = node;
        };
    }
    componentDidMount() {
        const node = this.current;
        const { $component, $context, innerProps } = this.props;
        const { mount } = $component;
        if (node && (0, piral_base_1.isfunc)(mount)) {
            mount(node, innerProps, $context, this.local);
            node.addEventListener('render-html', this.handler, false);
        }
        this.previous = node;
    }
    componentDidUpdate() {
        const { current, previous } = this;
        const { $component, $context, innerProps } = this.props;
        const { update } = $component;
        if (current !== previous) {
            previous && this.componentWillUnmount();
            current && this.componentDidMount();
        }
        else if ((0, piral_base_1.isfunc)(update)) {
            update(current, innerProps, $context, this.local);
        }
    }
    componentWillUnmount() {
        const node = this.previous;
        const { $component } = this.props;
        const { unmount } = $component;
        if (node && (0, piral_base_1.isfunc)(unmount)) {
            unmount(node, this.local);
            node.removeEventListener('render-html', this.handler, false);
        }
        this.previous = undefined;
    }
    render() {
        const { $portalId } = this.props;
        return React.createElement("div", { "data-portal-id": $portalId, ref: this.setNode });
    }
}
exports.ForeignComponentContainer = ForeignComponentContainer;
