// piral-core/lib/state/createGlobalState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGlobalState = void 0;
const react_atom_1 = require("@dbeining/react-atom");
const components_1 = require("../components");
function extend(defaultState, customState) {
    for (const key of Object.keys(customState)) {
        if (key === '__proto__' || key === 'constructor') {
            continue;
        }
        const value = customState[key];
        const original = defaultState[key];
        const nested = typeof original === 'object' && typeof value === 'object';
        defaultState[key] = nested ? extend(original, value) : value;
    }
    return defaultState;
}
function createGlobalState(customState = {}) {
    const defaultState = {
        app: {
            error: undefined,
            loading: typeof window !== 'undefined',
            layout: 'desktop',
            publicPath: process.env.PIRAL_PUBLIC_PATH || '/',
        },
        components: {
            ErrorInfo: components_1.DefaultErrorInfo,
            LoadingIndicator: components_1.DefaultLoadingIndicator,
            Router: components_1.DefaultRouter,
            RouteSwitch: components_1.DefaultRouteSwitch,
            Layout: components_1.DefaultLayout,
        },
        errorComponents: {},
        registry: {
            extensions: {},
            pages: {},
            wrappers: {},
        },
        routes: {},
        data: {},
        portals: {},
        modules: [],
    };
    return react_atom_1.Atom.of(extend(defaultState, customState));
}
exports.createGlobalState = createGlobalState;
