// piral-modals/lib/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModalsApi = void 0;
const actions = require("./actions");
const piral_core_1 = require("piral-core");
const default_1 = require("./default");
const Modals_1 = require("./Modals");
function getModalDialogs(dialogs) {
    const modals = {};
    for (const { name, component, defaults, layout = {} } of dialogs) {
        modals[`global-${name}`] = {
            pilet: undefined,
            name,
            component,
            defaults,
            layout,
        };
    }
    return modals;
}
function withModals(modals) {
    return (state) => (Object.assign(Object.assign({}, state), { components: Object.assign({ ModalsHost: default_1.DefaultHost, ModalsDialog: default_1.DefaultDialog }, state.components), registry: Object.assign(Object.assign({}, state.registry), { modals }), modals: [] }));
}
/**
 * Creates new Pilet API extensions for support modal dialogs.
 */
function createModalsApi(config = {}) {
    const { dialogs = [], selectId = (name) => `${name}-${~~(Math.random() * 10000)}` } = config;
    return (context) => {
        context.defineActions(actions);
        context.dispatch((0, piral_core_1.withAll)(withModals(getModalDialogs(dialogs)), (0, piral_core_1.withRootExtension)('piral-modals', Modals_1.Modals)));
        return (api, target) => {
            const pilet = target.name;
            return {
                showModal(simpleName, options) {
                    const name = (0, piral_core_1.buildName)(pilet, simpleName);
                    const dialog = {
                        id: selectId(name),
                        name,
                        alternative: simpleName,
                        options,
                        close() {
                            setTimeout(() => context.closeModal(dialog), 0);
                        },
                    };
                    context.openModal(dialog);
                    return dialog.close;
                },
                registerModal(name, arg, defaults, layout = {}) {
                    const id = (0, piral_core_1.buildName)(pilet, name);
                    context.registerModal(id, {
                        pilet,
                        name,
                        component: (0, piral_core_1.withApi)(context, arg, api, 'modal'),
                        defaults,
                        layout,
                    });
                    return () => api.unregisterModal(name);
                },
                unregisterModal(name) {
                    const id = (0, piral_core_1.buildName)(pilet, name);
                    context.unregisterModal(id);
                },
            };
        };
    };
}
exports.createModalsApi = createModalsApi;
