// piral-feeds/lib/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFeedOptions = void 0;
const piral_base_1 = require("piral-base");
const noop = () => { };
function createFeedOptions(id, resolver) {
    if ((0, piral_base_1.isfunc)(resolver)) {
        return {
            id,
            initialize() {
                return resolver();
            },
            connect() {
                return noop;
            },
            update(data) {
                return Promise.resolve(data);
            },
            immediately: false,
            reducers: {},
        };
    }
    else {
        return {
            id,
            initialize() {
                return resolver.initialize();
            },
            connect(cb) {
                if (typeof resolver.connect === 'function') {
                    return resolver.connect(cb);
                }
                else {
                    return noop;
                }
            },
            update(data, item) {
                if (typeof resolver.update === 'function') {
                    return resolver.update(data, item);
                }
                else {
                    return Promise.resolve(data);
                }
            },
            immediately: resolver.immediately,
            reducers: resolver.reducers || {},
        };
    }
}
exports.createFeedOptions = createFeedOptions;
