// piral-core/lib/modules/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCoreApi = void 0;
const element_1 = require("./element");
const state_1 = require("../state");
const components_1 = require("../components");
const utils_1 = require("../utils");
function createCoreApi(context) {
    return (api, meta) => {
        const pilet = meta.name;
        return {
            getData(name) {
                return context.readDataValue(name);
            },
            setData(name, value, options) {
                const { target = 'memory', expires } = (0, utils_1.createDataOptions)(options);
                const expiration = (0, utils_1.getDataExpiration)(expires);
                return context.tryWriteDataItem(name, value, pilet, target, expiration);
            },
            registerPage(route, arg, meta = {}) {
                const component = (0, state_1.withApi)(context, arg, api, 'page', undefined, { meta });
                context.registerPage(route, {
                    pilet,
                    meta,
                    component,
                });
                return () => api.unregisterPage(route);
            },
            unregisterPage(route) {
                context.unregisterPage(route);
            },
            registerExtension(name, arg, defaults) {
                context.registerExtension(name, {
                    pilet,
                    component: (0, state_1.withApi)(context, arg, api, 'extension'),
                    reference: arg,
                    defaults,
                });
                return () => api.unregisterExtension(name, arg);
            },
            unregisterExtension(name, arg) {
                context.unregisterExtension(name, arg);
            },
            renderHtmlExtension(element, props) {
                const [dispose] = (0, element_1.renderElement)(context, element, props);
                return dispose;
            },
            Extension: components_1.ExtensionSlot,
        };
    };
}
exports.createCoreApi = createCoreApi;
