// piral-core/esm/modules/element.js
import { ExtensionSlot } from '../components';
import { tryParseJson, noop, reactifyContent, renderInDom, changeDomPortal } from '../utils';
if (typeof window !== 'undefined' && 'customElements' in window) {
    class PiralExtension extends HTMLElement {
        constructor() {
            super(...arguments);
            this.dispose = noop;
            this.update = noop;
            this.props = {
                name: this.getAttribute('name'),
                params: tryParseJson(this.getAttribute('params')),
                empty: undefined,
                children: reactifyContent(this.childNodes),
            };
        }
        get params() {
            return this.props.params;
        }
        set params(value) {
            this.props.params = value;
            this.update(this.props);
        }
        get name() {
            return this.props.name;
        }
        set name(value) {
            this.props.name = value;
            this.update(this.props);
        }
        get empty() {
            return this.props.empty;
        }
        set empty(value) {
            this.props.empty = value;
            this.update(this.props);
        }
        connectedCallback() {
            if (this.isConnected) {
                this.dispatchEvent(new CustomEvent('render-html', {
                    bubbles: true,
                    detail: {
                        target: this,
                        props: this.props,
                    },
                }));
            }
        }
        disconnectedCallback() {
            this.dispose();
            this.dispose = noop;
            this.update = noop;
        }
        attributeChangedCallback(name, _, newValue) {
            switch (name) {
                case 'name':
                    this.name = newValue;
                    break;
                case 'params':
                    this.params = tryParseJson(newValue);
                    break;
            }
        }
        static get observedAttributes() {
            return ['name', 'params'];
        }
    }
    customElements.define('piral-extension', PiralExtension);
}
export function renderElement(context, element, props) {
    let [id, portal] = renderInDom(context, element, ExtensionSlot, props);
    const evName = 'extension-props-changed';
    const handler = (ev) => update(ev.detail);
    const dispose = () => {
        context.hidePortal(id, portal);
        element.removeEventListener(evName, handler);
    };
    const update = (newProps) => {
        [id, portal] = changeDomPortal(id, portal, context, element, ExtensionSlot, newProps);
    };
    element.addEventListener(evName, handler);
    return [dispose, update];
}
