// piral-ng/lib/NgExtension.js
"use strict";
var NgExtension_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NgExtension = void 0;
const tslib_1 = require("tslib");
const ngCore = require("@angular/core");
const core_1 = require("@angular/core");
const ngc = ngCore;
const selector = 'extension-component';
let NgExtension = NgExtension_1 = class NgExtension {
    constructor(elRef, piral) {
        this.elRef = elRef;
        this.piral = piral;
    }
    ngOnChanges() {
        this.elRef.nativeElement.dispatchEvent(new CustomEvent('extension-props-changed', {
            detail: {
                name: this.name,
                params: this.params,
            },
        }));
    }
    ngAfterContentInit() {
        this.piral.renderHtmlExtension(this.elRef.nativeElement, {
            name: this.name,
            params: this.params,
        });
    }
};
// @ts-ignore
NgExtension.ɵfac = 'ɵɵdirectiveInject' in ngc
    ? (t) => new (t || NgExtension_1)(ngc.ɵɵdirectiveInject(core_1.ElementRef), ngc.ɵɵdirectiveInject('piral'))
    : undefined;
// @ts-ignore
NgExtension.ɵcmp = 'ɵɵdefineComponent' in ngc
    ? ngc.ɵɵdefineComponent({
        type: NgExtension_1,
        selectors: [selector],
        inputs: { name: 'name', params: 'params' },
        decls: 0,
        vars: 0,
        template: () => { },
        encapsulation: 2,
    })
    : undefined;
(0, tslib_1.__decorate)([
    (0, core_1.Input)('name'),
    (0, tslib_1.__metadata)("design:type", String)
], NgExtension.prototype, "name", void 0);
(0, tslib_1.__decorate)([
    (0, core_1.Input)('params'),
    (0, tslib_1.__metadata)("design:type", Object)
], NgExtension.prototype, "params", void 0);
NgExtension = NgExtension_1 = (0, tslib_1.__decorate)([
    (0, core_1.Component)({
        selector,
        template: '',
    }),
    (0, tslib_1.__param)(1, (0, core_1.Inject)('piral')),
    (0, tslib_1.__metadata)("design:paramtypes", [core_1.ElementRef, Object])
], NgExtension);
exports.NgExtension = NgExtension;
if ('ɵsetClassMetadata' in ngc) {
    ngc.ɵsetClassMetadata(NgExtension, [
        {
            type: core_1.Component,
            args: [{ selector, template: '' }],
        },
    ]);
}
