// piral-ng/lib/RoutingService.js
"use strict";
var RoutingService_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoutingService = void 0;
const tslib_1 = require("tslib");
const ngCore = require("@angular/core");
const core_1 = require("@angular/core");
const router_1 = require("@angular/router");
const operators_1 = require("rxjs/operators");
const ngc = ngCore;
let RoutingService = RoutingService_1 = class RoutingService {
    constructor(context, router, zone) {
        this.context = context;
        this.router = router;
        this.zone = zone;
        this.invalidRoutes = [];
        if (this.router) {
            this.router.errorHandler = (error) => {
                if (error.message.match('Cannot match any routes')) {
                    // ignore this special error
                    return undefined;
                }
                throw error;
            };
            this.dispose = this.context.router.history.listen((e) => {
                const path = e.pathname;
                if (!this.invalidRoutes.includes(path)) {
                    this.zone.run(() => this.router.navigateByUrl(path));
                }
            });
            this.subscription = this.router.events.pipe((0, operators_1.filter)((e) => e instanceof router_1.NavigationError)).subscribe((e) => {
                const path = e.url;
                if (!this.invalidRoutes.includes(path)) {
                    this.invalidRoutes.push(path);
                }
                this.context.router.history.push(path);
            });
        }
    }
    ngOnDestroy() {
        var _a, _b;
        (_a = this.dispose) === null || _a === void 0 ? void 0 : _a.call(this);
        (_b = this.subscription) === null || _b === void 0 ? void 0 : _b.unsubscribe();
    }
};
RoutingService.ɵfac = 'ɵɵinject' in ngc
    ? (t) => new (t || RoutingService_1)(ngc.ɵɵinject('Context'), ngc.ɵɵinject(router_1.Router, 8), ngc.ɵɵinject(core_1.NgZone, 8))
    : undefined;
RoutingService.ɵprov = 'ɵɵngDeclareInjectable' in ngc
    ? ngc.ɵɵdefineInjectable({ token: RoutingService_1, factory: RoutingService_1.ɵfac })
    : undefined;
RoutingService = RoutingService_1 = (0, tslib_1.__decorate)([
    (0, core_1.Injectable)(),
    (0, tslib_1.__param)(0, (0, core_1.Inject)('Context')),
    (0, tslib_1.__param)(1, (0, core_1.Optional)()),
    (0, tslib_1.__param)(2, (0, core_1.Optional)()),
    (0, tslib_1.__metadata)("design:paramtypes", [Object, router_1.Router,
        core_1.NgZone])
], RoutingService);
exports.RoutingService = RoutingService;
if ('ɵsetClassMetadata' in ngc) {
    ngc.ɵsetClassMetadata(RoutingService, [
        {
            type: core_1.Injectable,
            args: [{ name: 'resourceUrl' }],
        },
    ], () => [
        {
            type: undefined,
            decorators: [
                {
                    type: core_1.Inject,
                    args: ['Context'],
                },
            ],
        },
        {
            type: router_1.Router,
            decorators: [
                {
                    type: core_1.Optional,
                },
            ],
        },
        {
            type: core_1.NgZone,
            decorators: [
                {
                    type: core_1.Optional,
                },
            ],
        },
    ]);
}
