// piral-ng/esm/ResourceUrlPipe.js
var ResourceUrlPipe_1;
import { __decorate, __metadata, __param } from "tslib";
import * as ngCore from '@angular/core';
import { Inject, Pipe } from '@angular/core';
const ngc = ngCore;
let ResourceUrlPipe = ResourceUrlPipe_1 = class ResourceUrlPipe {
    constructor(piral) {
        this.piral = piral;
    }
    transform(value) {
        const { basePath = '/' } = this.piral.meta;
        return basePath + value;
    }
};
// @ts-ignore
ResourceUrlPipe.ɵfac = 'ɵɵdirectiveInject' in ngc ? (t) => new (t || ResourceUrlPipe_1)(ngc.ɵɵdirectiveInject('piral', 16)) : undefined;
// @ts-ignore
ResourceUrlPipe.ɵpipe = 'ɵɵdefinePipe' in ngc ? ngc.ɵɵdefinePipe({ name: 'resourceUrl', type: ResourceUrlPipe_1, pure: true }) : undefined;
ResourceUrlPipe = ResourceUrlPipe_1 = __decorate([
    Pipe({ name: 'resourceUrl' }),
    __param(0, Inject('piral')),
    __metadata("design:paramtypes", [Object])
], ResourceUrlPipe);
export { ResourceUrlPipe };
if ('ɵsetClassMetadata' in ngc) {
    ngc.ɵsetClassMetadata(ResourceUrlPipe, [
        {
            type: Pipe,
            args: [{ name: 'resourceUrl' }],
        },
    ], () => [
        {
            type: undefined,
            decorators: [
                {
                    type: Inject,
                    args: ['piral'],
                },
            ],
        },
    ]);
}
