// piral-dashboard/esm/create.js
import * as actions from './actions';
import { buildName, withApi, withRootExtension, withAll } from 'piral-core';
import { DefaultTile, DefaultContainer } from './default';
import { Dashboard } from './Dashboard';
function getPreferences(defaultPreferences, customPreferences = {}) {
    return Object.assign(Object.assign({}, defaultPreferences), customPreferences);
}
function getTiles(items, defaultPreferences) {
    const tiles = {};
    let i = 0;
    for (const { component, preferences } of items) {
        tiles[`global-${i++}`] = {
            pilet: undefined,
            component,
            preferences: getPreferences(defaultPreferences, preferences),
        };
    }
    return tiles;
}
function withTiles(tiles) {
    return (state) => (Object.assign(Object.assign({}, state), { components: Object.assign({ DashboardTile: DefaultTile, DashboardContainer: DefaultContainer }, state.components), registry: Object.assign(Object.assign({}, state.registry), { tiles }) }));
}
/**
 * Creates the Pilet API extension for activating dashboard support.
 */
export function createDashboardApi(config = {}) {
    const { tiles = [], defaultPreferences = {} } = config;
    return (context) => {
        context.defineActions(actions);
        context.dispatch(withAll(withTiles(getTiles(tiles, defaultPreferences)), withRootExtension('piral-dashboard', Dashboard)));
        return (api, target) => {
            const pilet = target.name;
            let next = 0;
            return {
                registerTile(name, arg, preferences) {
                    if (typeof name !== 'string') {
                        preferences = arg;
                        arg = name;
                        name = next++;
                    }
                    const id = buildName(pilet, name);
                    context.registerTile(id, {
                        pilet,
                        component: withApi(context, arg, api, 'tile'),
                        preferences: getPreferences(defaultPreferences, preferences),
                    });
                    return () => api.unregisterTile(name);
                },
                unregisterTile(name) {
                    const id = buildName(pilet, name);
                    context.unregisterTile(id);
                },
            };
        };
    };
}
